src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/AutoInsert/RemoteAutoInsertService.cs

```diff
@@ -0,0 +1,112 @@
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor.Cohost.Handlers;
+using Microsoft.CodeAnalysis.Razor.AutoInsert;
+using Microsoft.CodeAnalysis.Razor.DocumentMapping;
+using Microsoft.CodeAnalysis.Razor.Protocol;
+using Microsoft.CodeAnalysis.Razor.Protocol.AutoInsert;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.CodeAnalysis.Remote.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+using Response = Microsoft.CodeAnalysis.Razor.Remote.RemoteResponse<Microsoft.CodeAnalysis.Razor.Protocol.AutoInsert.RemoteInsertTextEdit?>;
+using RoslynFormattingOptions = Roslyn.LanguageServer.Protocol.FormattingOptions;
+
+namespace Microsoft.CodeAnalysis.Remote.Razor;
+
+internal class RemoteAutoInsertService(in ServiceArgs args)
+    : RazorDocumentServiceBase(in args), IRemoteAutoInsertService
+{
+    internal sealed class Factory : FactoryBase<IRemoteAutoInsertService>
+    {
+        protected override IRemoteAutoInsertService CreateService(in ServiceArgs args)
+            => new RemoteAutoInsertService(in args);
+    }
+
+    private readonly IAutoInsertService _autoInsertService
+        = args.ExportProvider.GetExportedValue<IAutoInsertService>();
+    private readonly IDocumentMappingService _documentMappingService
+        = args.ExportProvider.GetExportedValue<IDocumentMappingService>();
+    private readonly IFilePathService _filePathService =
+        args.ExportProvider.GetExportedValue<IFilePathService>();
+
+    public ValueTask<Response> TryResolveInsertionAsync(
+        RazorPinnedSolutionInfoWrapper solutionInfo,
+        DocumentId documentId,
+        LinePosition linePosition,
+        string character,
+        bool autoCloseTags,
+        CancellationToken cancellationToken)
+        => RunServiceAsync(
+            solutionInfo,
+            documentId,
+            context => TryResolveInsertionAsync(
+                context,
+                linePosition,
+                character,
+                autoCloseTags,
+                cancellationToken),
+            cancellationToken);
+
+    private async ValueTask<Response> TryResolveInsertionAsync(
+        RemoteDocumentContext remoteDocumentContext,
+        LinePosition linePosition,
+        string character,
+        bool autoCloseTags,
+        CancellationToken cancellationToken)
+    {
+        var sourceText = await remoteDocumentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+        if (!sourceText.TryGetAbsoluteIndex(linePosition, out var index))
+        {
+            return Response.NoFurtherHandling;
+        }
+
+        var codeDocument = await remoteDocumentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+
+        var languageKind = _documentMappingService.GetLanguageKind(codeDocument, index, rightAssociative: true);
```

Comment:
Happened to be looking at the existing AutoInsert endpoint today, for other reasons, and noticed is uses this: https://github.com/dotnet/razor/blob/main/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AutoInsert/PreferHtmlInAttributeValuesDocumentPositionStrategy.cs#L19

We'll need to do that here too, to get matching behaviour. Should be able to just move it down to Workspaces, and call the static Instance property etc.